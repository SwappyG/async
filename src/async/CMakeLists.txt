cmake_minimum_required(VERSION 3.0)
project(async)

enable_testing()

set(CMAKE_CXX_STANDARD 17)

set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
set(THREADS_PREFER_PTHREAD_FLAG TRUE)
set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} "-pthread")

find_package(Boost REQUIRED COMPONENTS unit_test_framework)
find_package(Threads REQUIRED)

include_directories (${Boost_INCLUDE_DIRS})

set(INC
  inc/${PROJECT_NAME}/atomic.hpp
  inc/${PROJECT_NAME}/atomic_base.hpp
  inc/${PROJECT_NAME}/atomic_proxies.hpp
  inc/${PROJECT_NAME}/atomic_rw.hpp
  inc/${PROJECT_NAME}/atomic_rw_base.hpp
  inc/${PROJECT_NAME}/optional.hpp
  inc/${PROJECT_NAME}/optional_base.hpp
)

add_library(${PROJECT_NAME} INTERFACE)
target_include_directories(${PROJECT_NAME} INTERFACE inc/)
target_link_libraries(${PROJECT_NAME} INTERFACE meta Threads::Threads)

install(TARGETS ${PROJECT_NAME} DESTINATION lib)
install(FILES ${INC} DESTINATION inc)

function(add_boost_test test_name)
  add_executable(${test_name} tests/${test_name}.cpp)
  target_link_libraries(${test_name} ${Boost_LIBRARIES} ${PROJECT_NAME} ${ARGN})
  add_test(${test_name}_test COMMAND ${test_name})
endfunction()

add_boost_test(atomic_test)
add_boost_test(atomic_rw_test)
add_boost_test(container_test)
add_boost_test(container_traits_test)
add_boost_test(optional_test)
