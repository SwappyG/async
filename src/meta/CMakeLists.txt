cmake_minimum_required(VERSION 3.0)
project(meta)

enable_testing()

set(CMAKE_CXX_STANDARD 17)

find_package(Boost REQUIRED COMPONENTS unit_test_framework)

include_directories (${Boost_INCLUDE_DIRS})

include_directories (${CMAKE_SOURCE_DIR})

set(INC
  inc/${PROJECT_NAME}/detector.hpp
  inc/${PROJECT_NAME}/devoid.hpp
  inc/${PROJECT_NAME}/enable_if.hpp
  inc/${PROJECT_NAME}/identity.hpp
  inc/${PROJECT_NAME}/none_such.hpp
  inc/${PROJECT_NAME}/void_safe.hpp
)

add_library(${PROJECT_NAME} INTERFACE)
target_include_directories(${PROJECT_NAME} INTERFACE inc/)
target_link_libraries(${PROJECT_NAME} INTERFACE ${Boost_LIBRARIES})

install(
  TARGETS ${PROJECT_NAME} DESTINATION lib
)

install(FILES ${INC} DESTINATION inc)