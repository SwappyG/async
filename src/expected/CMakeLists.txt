cmake_minimum_required(VERSION 3.0)
project(expected)
enable_testing()

set(CMAKE_CXX_STANDARD 17)

find_package(Boost REQUIRED COMPONENTS unit_test_framework)
find_package(nil_cpp COMPONENTS meta)

include_directories (${Boost_INCLUDE_DIRS})

set(INC
  inc/${PROJECT_NAME}/expected.hpp
  inc/${PROJECT_NAME}/expected_traits.hpp
  inc/${PROJECT_NAME}/unexpected.hpp
  inc/${PROJECT_NAME}/unexpected_traits.hpp
)

add_library(${PROJECT_NAME} INTERFACE)
target_include_directories(${PROJECT_NAME} INTERFACE inc/)
target_link_libraries(${PROJECT_NAME} INTERFACE meta)

install(TARGETS ${PROJECT_NAME} DESTINATION lib)
install(FILES ${INC} DESTINATION inc)

function(add_boost_test test_name)
  message(STATUS argn: ${ARGN})
  message(STATUS ${test_name})
  add_executable(${test_name} tests/${test_name}.cpp)
  target_link_libraries(${test_name} ${Boost_LIBRARIES} ${PROJECT_NAME} ${ARGN})
  add_test(${test_name}_test COMMAND ${test_name})
endfunction()

add_boost_test(unexpected_test meta)
add_boost_test(expected_test meta)